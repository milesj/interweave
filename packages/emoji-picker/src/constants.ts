/* eslint-disable sort-keys */

import {
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_COMPONENT,
	GROUP_KEY_FLAGS,
	GROUP_KEY_FOOD_DRINK,
	GROUP_KEY_OBJECTS,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_SMILEYS_EMOTION,
	GROUP_KEY_SYMBOLS,
	GROUP_KEY_TRAVEL_PLACES,
	SKIN_KEY_DARK,
	SKIN_KEY_LIGHT,
	SKIN_KEY_MEDIUM,
	SKIN_KEY_MEDIUM_DARK,
	SKIN_KEY_MEDIUM_LIGHT,
} from 'emojibase';
import { GroupKey, SkinToneKey } from './types';

export {
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_COMPONENT,
	GROUP_KEY_FLAGS,
	GROUP_KEY_FOOD_DRINK,
	GROUP_KEY_OBJECTS,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_SMILEYS_EMOTION,
	GROUP_KEY_SYMBOLS,
	GROUP_KEY_TRAVEL_PLACES,
	SKIN_KEY_DARK,
	SKIN_KEY_LIGHT,
	SKIN_KEY_MEDIUM,
	SKIN_KEY_MEDIUM_DARK,
	SKIN_KEY_MEDIUM_LIGHT,
};

export const GROUP_KEY_COMMONLY_USED = 'commonly-used';
export const GROUP_KEY_SEARCH_RESULTS = 'search-results';
export const GROUP_KEY_VARIATIONS = 'variations';
export const GROUP_KEY_NONE = 'none';

export const GROUPS: GroupKey[] = [
	GROUP_KEY_SMILEYS_EMOTION,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_COMPONENT, // Unused but required for order
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_FOOD_DRINK,
	GROUP_KEY_TRAVEL_PLACES,
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_OBJECTS,
	GROUP_KEY_SYMBOLS,
	GROUP_KEY_FLAGS,
];

export const GROUP_ICONS: Record<string, string> = {
	[GROUP_KEY_COMMONLY_USED]: 'üïë',
	[GROUP_KEY_SMILEYS_EMOTION]: 'üòÉ',
	[GROUP_KEY_PEOPLE_BODY]: 'üëç',
	[GROUP_KEY_ANIMALS_NATURE]: 'üåø',
	[GROUP_KEY_FOOD_DRINK]: 'üçé',
	[GROUP_KEY_TRAVEL_PLACES]: 'üó∫Ô∏è',
	[GROUP_KEY_ACTIVITIES]: '‚öΩÔ∏è',
	[GROUP_KEY_OBJECTS]: 'üìò',
	[GROUP_KEY_SYMBOLS]: '‚õîÔ∏è',
	[GROUP_KEY_FLAGS]: 'üè¥',
};

export const SKIN_KEY_NONE = 'none';

export const SKIN_TONES: SkinToneKey[] = [
	SKIN_KEY_NONE,
	SKIN_KEY_LIGHT,
	SKIN_KEY_MEDIUM_LIGHT,
	SKIN_KEY_MEDIUM,
	SKIN_KEY_MEDIUM_DARK,
	SKIN_KEY_DARK,
];

export const SKIN_COLORS = {
	[SKIN_KEY_NONE]: '#FFCC22',
	[SKIN_KEY_LIGHT]: '#FADCBC',
	[SKIN_KEY_MEDIUM_LIGHT]: '#E0BB95',
	[SKIN_KEY_MEDIUM]: '#BF8F68',
	[SKIN_KEY_MEDIUM_DARK]: '#9B643D',
	[SKIN_KEY_DARK]: '#5A463A',
};

export const SCROLL_BUFFER = 150;
export const SCROLL_DEBOUNCE = 100;
export const SEARCH_THROTTLE = 300;

export const KEY_COMMONLY_USED = 'interweave/emoji/commonlyUsed';
export const KEY_SKIN_TONE = 'interweave/emoji/skinTone';

export const COMMON_MODE_RECENT = 'recently-used';
export const COMMON_MODE_FREQUENT = 'frequently-used';

export const CONTEXT_CLASSNAMES = {
	picker: 'interweave-picker__picker',
	emoji: 'interweave-picker__emoji',
	emojiActive: 'interweave-picker__emoji--active',
	emojis: 'interweave-picker__emojis',
	emojisList: 'interweave-picker__emojis-list',
	emojisRow: 'interweave-picker__emojis-row',
	emojisHeader: 'interweave-picker__emojis-header',
	emojisHeaderSticky: 'interweave-picker__emojis-header--sticky',
	emojisBody: 'interweave-picker__emojis-body',
	group: 'interweave-picker__group',
	groupActive: 'interweave-picker__group--active',
	groups: 'interweave-picker__groups',
	groupsList: 'interweave-picker__groups-list',
	skinTone: 'interweave-picker__skin-tone',
	skinToneActive: 'interweave-picker__skin-tone--active',
	skinTones: 'interweave-picker__skin-tones',
	skinTonesList: 'interweave-picker__skin-tones-list',
	noPreview: 'interweave-picker__no-preview',
	noResults: 'interweave-picker__no-results',
	preview: 'interweave-picker__preview',
	previewEmoji: 'interweave-picker__preview-emoji',
	previewContent: 'interweave-picker__preview-content',
	previewTitle: 'interweave-picker__preview-title',
	previewSubtitle: 'interweave-picker__preview-subtitle',
	previewShiftMore: 'interweave-picker__preview-more',
	search: 'interweave-picker__search',
	searchInput: 'interweave-picker__search-input',
	clear: 'interweave-picker__clear',
};

export const CONTEXT_MESSAGES = {
	frequentlyUsed: 'Frequently used',
	recentlyUsed: 'Recently used',
	variations: 'Variations',
	none: 'All emojis',
	// Skin tones
	skinNone: 'No skin tone',
	skinLight: 'Light skin tone',
	skinMediumLight: 'Medium-light skin tone',
	skinMedium: 'Medium skin tone',
	skinMediumDark: 'Medium-dark skin tone',
	skinDark: 'Dark skin tone',
	// Miscellaneous
	search: 'Search',
	searchA11y: 'Search for emojis by keyword',
	searchResults: 'Search results',
	noPreview: '',
	noResults: 'No results',
	clearUsed: 'Clear used',
};
