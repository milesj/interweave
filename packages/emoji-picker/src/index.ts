/**
 * @copyright   2016-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Picker } from './Picker';

export * from './constants';
export * from './types';

export default Picker;
