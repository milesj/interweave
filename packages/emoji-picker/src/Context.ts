import React from 'react';
import { EmojiDataManager, LATEST_DATASET_VERSION } from 'interweave-emoji';
import { CONTEXT_CLASSNAMES, CONTEXT_MESSAGES } from './constants';
import { Context as ContextType } from './types';

const version = process.env.NODE_ENV === 'test' ? '0.0.0' : LATEST_DATASET_VERSION;

export const Context = React.createContext<ContextType>({
	classNames: CONTEXT_CLASSNAMES,
	emojiData: EmojiDataManager.getInstance('en', version),
	emojiLargeSize: 0,
	emojiPadding: 0,
	emojiPath: '{{hexcode}}',
	emojiSize: 0,
	emojiSource: {
		compact: false,
		locale: 'en',
		version,
	},
	messages: CONTEXT_MESSAGES,
});
