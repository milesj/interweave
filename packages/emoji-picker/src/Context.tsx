import React from 'react';
import { EmojiDataManager, LATEST_DATASET_VERSION } from 'interweave-emoji';
import { CONTEXT_CLASSNAMES, CONTEXT_MESSAGES } from './constants';
import { Context } from './types';

export default React.createContext<Context>({
	classNames: CONTEXT_CLASSNAMES,
	emojiData: EmojiDataManager.getInstance('en'),
	emojiLargeSize: 0,
	emojiPadding: 0,
	emojiPath: '{{hexcode}}',
	emojiSize: 0,
	emojiSource: {
		compact: false,
		locale: 'en',
		version: LATEST_DATASET_VERSION,
	},
	messages: CONTEXT_MESSAGES,
});
