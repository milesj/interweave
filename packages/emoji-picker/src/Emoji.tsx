import React, { useContext } from 'react';
import { Emoji as EmojiCharacter, CanonicalEmoji } from 'interweave-emoji';
import Context from './Context';

export interface EmojiProps {
  active: boolean;
  emoji: CanonicalEmoji;
  onEnter: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onLeave: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onSelect: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
}

export default function Emoji({ active, emoji, onEnter, onLeave, onSelect }: EmojiProps) {
  const { classNames, emojiPadding, emojiPath, emojiSize, emojiSource } = useContext(Context);
  const dimension = emojiPadding + emojiPadding + emojiSize;
  const className = [classNames.emoji];

  if (active) {
    className.push(classNames.emojiActive);
  }

  // Handlers
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    onSelect(emoji, event);
  };

  const handleEnter = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    onEnter(emoji, event);
  };

  const handleLeave = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    onLeave(emoji, event);
  };

  return (
    <button
      key={emoji.hexcode}
      className={className.join(' ')}
      style={{
        height: dimension,
        padding: emojiPadding,
        width: dimension,
      }}
      title={emoji.annotation}
      type="button"
      onClick={handleClick}
      onMouseEnter={handleEnter}
      onMouseLeave={handleLeave}
    >
      <EmojiCharacter
        path={emojiPath}
        size={emojiSize}
        source={emojiSource}
        hexcode={emoji.hexcode}
      />
    </button>
  );
}
