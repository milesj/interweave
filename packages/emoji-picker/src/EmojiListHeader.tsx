import React, { useCallback, useContext } from 'react';
import {
	GROUP_KEY_COMMONLY_USED,
	GROUP_KEY_NONE,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_SEARCH_RESULTS,
	GROUP_KEY_VARIATIONS,
} from './constants';
import { Context } from './Context';
import { useGroupMessage } from './hooks/useGroupMessage';
import { CommonMode, GroupKey } from './types';

export interface EmojiListHeaderProps {
	clearIcon?: React.ReactNode;
	commonMode: CommonMode;
	group: GroupKey;
	onClear: () => void;
	skinTonePalette?: React.ReactNode;
	sticky?: boolean;
}

export function EmojiListHeader({
	clearIcon,
	commonMode,
	group,
	skinTonePalette,
	sticky,
	onClear,
}: EmojiListHeaderProps) {
	const { classNames, messages } = useContext(Context);
	const showClear =
		clearIcon && (group === GROUP_KEY_COMMONLY_USED || group === GROUP_KEY_VARIATIONS);
	const showPalette =
		skinTonePalette &&
		(group === GROUP_KEY_PEOPLE_BODY ||
			group === GROUP_KEY_SEARCH_RESULTS ||
			group === GROUP_KEY_NONE);
	const className = [classNames.emojisHeader];
	const title = useGroupMessage(group, commonMode);

	if (sticky) {
		className.push(classNames.emojisHeaderSticky);
	}

	const handleClear = useCallback(
		(event: React.MouseEvent<HTMLButtonElement>) => {
			event.preventDefault();
			onClear();
		},
		[onClear],
	);

	return (
		<header className={className.join(' ')}>
			<span>{title}</span>

			{showPalette && skinTonePalette}

			{showClear && (
				<button
					className={classNames.clear}
					title={messages.clearUsed}
					type="button"
					onClick={handleClear}
				>
					{clearIcon}
				</button>
			)}
		</header>
	);
}
