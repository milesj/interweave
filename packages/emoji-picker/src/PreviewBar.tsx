import React, { useContext } from 'react';
import { Emoji as EmojiCharacter, CanonicalEmoji } from 'interweave-emoji';
import Context from './Context';

const TITLE_REGEX = /(^|:|\.)\s?[a-z]/g;

export interface PreviewBarProps {
  emoji?: CanonicalEmoji | null;
  hideEmoticon: boolean;
  hideShortcodes: boolean;
  noPreview?: React.ReactNode;
}

// Format the title using sentence case.
function formatTitle(title: string): string {
  return title.replace(TITLE_REGEX, token => token.toUpperCase());
}

// eslint-disable-next-line complexity
export default function PreviewBar({
  emoji,
  hideEmoticon,
  hideShortcodes,
  noPreview,
}: PreviewBarProps) {
  const { classNames, emojiLargeSize, emojiPath, emojiSource, messages } = useContext(Context);

  if (!emoji) {
    const preview = noPreview || messages.noPreview;

    return (
      <section className={classNames.preview}>
        {preview && <div className={classNames.noPreview}>{preview}</div>}
      </section>
    );
  }

  const title = emoji.annotation || emoji.name;
  const subtitle = [];

  if (!hideEmoticon && emoji.emoticon) {
    subtitle.push(emoji.emoticon);
  }

  if (!hideShortcodes && emoji.canonical_shortcodes) {
    subtitle.push(...emoji.canonical_shortcodes);
  }

  return (
    <section className={classNames.preview}>
      <div className={classNames.previewEmoji}>
        <EmojiCharacter
          enlarged
          hexcode={emoji.hexcode}
          largeSize={emojiLargeSize}
          path={emojiPath}
          source={emojiSource}
        />
      </div>

      <div className={classNames.previewContent}>
        {title && (
          <div className={classNames.previewTitle}>
            {formatTitle(title)}

            {emoji.skins && emoji.skins.length > 0 && (
              <span className={classNames.previewShiftMore}>(+{emoji.skins.length})</span>
            )}
          </div>
        )}

        {subtitle.length > 0 && (
          <div className={classNames.previewSubtitle}>{subtitle.join(' ')}</div>
        )}
      </div>
    </section>
  );
}
