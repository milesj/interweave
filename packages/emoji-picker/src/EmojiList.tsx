import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { FixedSizeList as List, ListOnItemsRenderedProps } from 'react-window';
import chunk from 'lodash/chunk';
import { GROUP_KEY_COMPONENT, GROUP_KEY_NONE } from './constants';
import Context from './Context';
import EmojiListHeader from './EmojiListHeader';
import EmojiListRow, { EmojiListRowProps, VirtualRow } from './EmojiListRow';
import { GroupEmojiMap, GroupKey } from './types';

export interface IndicesMap {
	[key: string]: number;
}

export interface EmojiListProps extends EmojiListRowProps {
	activeGroup: GroupKey;
	columnCount: number;
	columnPadding?: number;
	groupedEmojis: GroupEmojiMap;
	hideGroupHeaders: boolean;
	noResults?: React.ReactNode;
	onScroll: () => void;
	onScrollGroup: (group: GroupKey) => void;
	rowCount: number;
	rowPadding?: number;
	scrollToGroup?: GroupKey | '';
	stickyGroupHeader?: boolean;
}

export default function EmojiList({
	activeGroup,
	columnCount,
	columnPadding = 0,
	groupedEmojis,
	hideGroupHeaders,
	noResults,
	rowCount,
	rowPadding = 0,
	scrollToGroup,
	stickyGroupHeader,
	onScroll,
	onScrollGroup,
	...rowProps
}: EmojiListProps) {
	const { classNames, emojiPadding, emojiSize, messages } = useContext(Context);
	const [rows, setRows] = useState<VirtualRow[]>([]);
	const [indices, setIndices] = useState<IndicesMap>({});
	const ref = useRef<List>(null);
	const size = emojiSize + emojiPadding * 2;
	const rowHeight = size + rowPadding * 2;
	const columnWidth = size + columnPadding * 2;

	// When emojis or virtual list props change,
	// we need to regenerate the list of rows.
	useEffect(() => {
		const virtualRows: VirtualRow[] = [];
		const nextIndices: IndicesMap = {
			'': -1, // Handle empty scroll to's
		};

		Object.keys(groupedEmojis).forEach((group) => {
			nextIndices[group] = virtualRows.length;

			if (group === GROUP_KEY_COMPONENT) {
				return;
			}

			if (!hideGroupHeaders) {
				virtualRows.push(group);
			}

			virtualRows.push(...chunk(groupedEmojis[group].emojis, columnCount));
		});

		setRows(virtualRows);
		setIndices(nextIndices);
	}, [columnCount, groupedEmojis, hideGroupHeaders]);

	// Scroll to the defined group when all data is available
	useEffect(() => {
		if (ref.current && scrollToGroup && indices[scrollToGroup] >= 0) {
			ref.current.scrollToItem(indices[scrollToGroup], 'start');
		}
	}, [scrollToGroup, indices]);

	// Loop over each group section within the scrollable container
	// and determine the active group.
	const handleRendered = useCallback(
		({ visibleStartIndex }: ListOnItemsRenderedProps) => {
			let lastGroup = '';

			Object.keys(indices).some((group) => {
				const index = indices[group];

				// Special case for commonly used and smileys, as they usually both render in the same view
				if (index === 0 && visibleStartIndex === 0) {
					lastGroup = group;

					return true;
				}

				// When we have to sticky headers, we need to change the header on the index right
				// before the next header, otherwise the change will happen too late
				if (stickyGroupHeader && index >= visibleStartIndex + 1) {
					return true;
					// Otherwise, we should update the active group when half way through the list
				} else if (!stickyGroupHeader && index >= visibleStartIndex + rowCount / 2) {
					return true;
				}

				lastGroup = group;

				return false;
			});

			// Only update if a different group
			if (lastGroup && lastGroup !== activeGroup) {
				onScrollGroup(lastGroup as GroupKey);
			}
		},
		[activeGroup, indices, onScrollGroup, rowCount, stickyGroupHeader],
	);

	// If no items to display, just return null
	if (rows.length === 0) {
		return <div className={classNames.noResults}>{noResults || messages.noResults}</div>;
	}

	return (
		<div className={classNames.emojis}>
			<List
				ref={ref}
				className={classNames.emojisList}
				height={rowHeight * rowCount}
				itemCount={rows.length}
				itemData={rows}
				itemSize={rowHeight}
				overscanCount={rowCount / 2}
				width={columnWidth * columnCount}
				onItemsRendered={handleRendered}
				onScroll={onScroll}
			>
				{(props) => <EmojiListRow {...rowProps} {...props} />}
			</List>

			{stickyGroupHeader && activeGroup !== GROUP_KEY_NONE && (
				<EmojiListHeader {...rowProps} sticky group={activeGroup} />
			)}
		</div>
	);
}
