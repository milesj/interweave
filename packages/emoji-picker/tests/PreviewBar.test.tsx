import React from 'react';
import { Emoji, EmojiDataManager } from 'interweave-emoji';
import { render } from 'rut-dom';
import { PreviewBar, PreviewBarProps } from '../src/PreviewBar';
import { CAT_EMOJI, ContextWrapper } from './mocks';

describe('PreviewBar', () => {
	const props: PreviewBarProps = {
		emoji: CAT_EMOJI,
		hideEmoticon: false,
		hideShortcodes: false,
	};

	beforeEach(() => {
		props.emoji = EmojiDataManager.getInstance('en').EMOJIS['1F642'];
	});

	it('renders a preview', () => {
		const { root } = render<PreviewBarProps>(<PreviewBar {...props} />, {
			wrapper: <ContextWrapper />,
		});

		expect(root.findOne('section')).toMatchSnapshot();
	});

	describe('with emoji', () => {
		it('can customize class name', () => {
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} />, {
				wrapper: (
					<ContextWrapper
						classNames={{
							preview: 'test-preview',
							previewEmoji: 'test-preview-emoji',
							previewContent: 'test-preview-content',
							previewTitle: 'test-preview-title',
							previewSubtitle: 'test-preview-subtitle',
						}}
					/>
				),
			});
			const divs = root.find('div');

			expect(root.findOne('section')).toHaveProp('className', 'test-preview');
			expect(divs[0]).toHaveProp('className', 'test-preview-emoji');
			expect(divs[1]).toHaveProp('className', 'test-preview-content');
			expect(divs[2]).toHaveProp('className', 'test-preview-title');
			expect(divs[3]).toHaveProp('className', 'test-preview-subtitle');
		});

		it('renders an emoji', () => {
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} />);

			expect(root.findOne(Emoji)).toHaveProps(
				expect.objectContaining({
					enlargeEmoji: true,
					hexcode: props.emoji!.hexcode,
				}),
			);
		});

		it('displays title', () => {
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} />);

			expect(root).toContainNode('Slightly smiling face');
		});

		it('doesnt display title if no annotation or name', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar
					{...props}
					emoji={{
						...props.emoji!,
						annotation: '',
					}}
				/>,
			);

			expect(root.find('div', { className: 'interweave-picker__preview-title' })).toHaveLength(0);
		});

		it('displays subtitle', () => {
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} />);

			expect(root).toContainNode(':) :slightly_smiling_face:');
		});

		it('displays subtitle without emoticon', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar
					{...props}
					emoji={{
						...props.emoji!,
						emoticon: '',
					}}
				/>,
			);

			expect(root).toContainNode(':slightly_smiling_face:');
		});

		it('displays subtitle without shortcodes', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar
					{...props}
					emoji={{
						...props.emoji!,
						canonical_shortcodes: [],
					}}
				/>,
			);

			expect(root).toContainNode(':)');
		});

		it('hides emoticon in subtitle', () => {
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} hideEmoticon />);

			expect(root).toContainNode(':slightly_smiling_face:');
		});

		it('hides shortcodes in subtitle', () => {
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} hideShortcodes />);

			expect(root).toContainNode(':)');
		});

		it('doesnt display subtitle if no emoticon or shortcodes', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar
					{...props}
					emoji={{
						...props.emoji!,
						canonical_shortcodes: [],
						emoticon: '',
					}}
				/>,
			);

			expect(root.find('div', { className: 'interweave-picker__preview-subtitle' })).toHaveLength(
				0,
			);
		});

		it('doesnt display subtitle if all are hidden', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar {...props} hideEmoticon hideShortcodes />,
			);

			expect(root.find('div', { className: 'interweave-picker__preview-subtitle' })).toHaveLength(
				0,
			);
		});
	});

	describe('no emoji', () => {
		it('can customize class name', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar {...props} emoji={null} noPreview="Preview" />,
				{
					wrapper: (
						<ContextWrapper
							classNames={{
								preview: 'test-preview',
								noPreview: 'test-preview-none',
							}}
						/>
					),
				},
			);

			expect(root.findOne('section')).toHaveProp('className', 'test-preview');
			expect(root.findOne('div')).toHaveProp('className', 'test-preview-none');
		});

		it('renders no preview message', () => {
			const none = 'No preview';
			const { root } = render<PreviewBarProps>(<PreviewBar {...props} emoji={null} />, {
				wrapper: (
					<ContextWrapper
						messages={{
							noPreview: none,
						}}
					/>
				),
			});

			expect(root).toContainNode(none);
		});

		it('renders no preview element', () => {
			const none = <b>No preview</b>;
			const { root } = render<PreviewBarProps>(
				<PreviewBar {...props} emoji={null} noPreview={none} />,
			);

			expect(root).toContainNode(none);
		});

		it('renders nothing if no preview message or preview element', () => {
			const { root } = render<PreviewBarProps>(
				<PreviewBar {...props} emoji={null} noPreview={null} />,
				{
					wrapper: (
						<ContextWrapper
							messages={{
								noPreview: '',
							}}
						/>
					),
				},
			);

			expect(root.findOne('section')).toMatchSnapshot();
		});
	});
});
