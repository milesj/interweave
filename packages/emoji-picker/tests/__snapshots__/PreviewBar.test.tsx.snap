// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewBar no emoji renders nothing if no preview message or preview element 1`] = `
<section className="interweave-picker__preview">
  
</section>
`;

exports[`PreviewBar renders a preview 1`] = `
<section className="interweave-picker__preview">
  <div className="interweave-picker__preview-emoji">
    <Emoji
      enlargeEmoji
      emojiLargeSize={72}
      emojiPath="{{hexcode}}"
      emojiSource={{ compact: false, locale: "en", version: "0.0.0" }}
      hexcode="1F642"
    >
      <img
        alt="ðŸ™‚"
        aria-label="slightly smiling face"
        data-emoticon=":)"
        data-hexcode="1F642"
        data-shortcodes=":slightly_smiling_face:"
        data-unicode="ðŸ™‚"
        src="1F642"
        style={{ display: "inline-block", verticalAlign: "middle", width: 72, height: 72 }}
        title="slightly smiling face"
      />
    </Emoji>
  </div>
  <div className="interweave-picker__preview-content">
    <div className="interweave-picker__preview-title">
      Slightly smiling face
    </div>
    <div className="interweave-picker__preview-subtitle">
      :) :slightly_smiling_face:
    </div>
  </div>
</section>
`;
